java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java

```diff
@@ -45,18 +51,61 @@ public void visitNode(Tree tree) {
     SyntaxToken identifierToken = variableTree.simpleName().identifierToken();
     while (type.is(Tree.Kind.ARRAY_TYPE)) {
       ArrayTypeTree arrayTypeTree = (ArrayTypeTree) type;
-      SyntaxToken arrayDesignatorToken = arrayTypeTree.ellipsisToken();
-      if (arrayDesignatorToken == null) {
-        arrayDesignatorToken = arrayTypeTree.openBracketToken();
-      }
-      if (isInvalidPosition(arrayDesignatorToken, identifierToken)) {
-        reportIssue(arrayDesignatorToken, "Move the array designator from the variable to the type.");
+      SyntaxToken openBracketToken = arrayTypeTree.openBracketToken();
+      if (openBracketToken != null && isInvalidPosition(openBracketToken, identifierToken)) {
+        ((InternalJavaIssueBuilder) ((DefaultJavaFileScannerContext) context).newIssue())
+          .forRule(this)
+          .onTree(openBracketToken)
+          .withMessage("Move the array designator from the variable to the type.")
+          .withQuickFixes(() -> createQuickFixes(variableTree, arrayTypeTree))
+          .report();
         break;
       }
       type = arrayTypeTree.type();
     }
   }
 
+  private static List<JavaQuickFix> createQuickFixes(VariableTree variableTree, ArrayTypeTree arrayTypeTree) {
+    if (isDeclarationTypeUsedBySeveralVariable(variableTree)) {
+      return Collections.emptyList();
+    }
+    return Collections.singletonList(
+      JavaQuickFix.newQuickFix("Move [] to the variable type")
+        .addTextEdit(JavaTextEdit.replaceBetweenTree(
+          arrayTypeTree.openBracketToken(),
+          arrayTypeTree.closeBracketToken(),
+          ""))
+        .addTextEdit(JavaTextEdit.insertAfterTree(
+          arrayTypeTree.type(),
+          "[]"))
+        .build());
+  }
+
+  private static boolean isDeclarationTypeUsedBySeveralVariable(VariableTree current) {
+    Tree parent = current.parent();
```

Comment:
I would cast the parent into a `JavaTree`, then you can call `children()`

Reply:
It makes sense for `BlockTree` and Tree.Kind.LIST but for `ClassTree` it's not efficient, instead of directly iterating the `members` you will previously iterate:
```
modifiers, atToken, declarationKeyword, simpleName, typeParameters..., recordComponents..., extendsKeyword,
superClass, implementsKeyword, superInterfaces..., permittedTypes, openBraceToken, and finally members
```
Same for `MethodTree` with `modifiers, typeParameters, returnType, simpleName, openParenToken, and finally members`